### PR TITLE
Add ROUGE Metric

### PR DESCRIPTION
**Pull request by go-with-me000**
Description：增加rouge metric

Main reason: 
增加rouge metric，修改bleu，并且修改element的数组无法进行数学运算的bug

Checklist  检查下面各项是否完成

Please feel free to remove inapplicable items for your PR.

-	[y] The PR title starts with [$CATEGORY] (例如[bugfix]修复bug，[new]添加新功能，[test]修改测试，[rm]删除旧代码)
-	[y] Changes are complete (i.e. I finished coding on this PR)  修改完成才提PR
-	[y] All changes have test coverage  修改的部分顺利通过测试。对于fastnlp/fastnlp/*的修改，测试代码**必须**提供在fastnlp/test/*。
-	[y] Code is well-documented  注释写好，API文档会从注释中抽取
-	[y] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change  修改导致例子或tutorial有变化，请找核心开发人员

Changes: 逐项描述修改的内容
- 增加rouge以及测试
- 修改bleu
- 修改element的数组无法进行数学运算的bug

Mention: 找人review你的PR

@修改过这个文件的人
@核心开发人员

### CODE REVIEW

File: fastNLP/core/metrics/bleu.py [hunk 1]
```diff
@@ -25,68 +22,88 @@ def tokenizer_fn(sentence: str) -> Sequence[str]:
...
 
         for prediction, references in zip(predictions_token, references_token):
             self.pred_len += len(prediction)
-            references_len_list = [len(reference) for reference in references]
-            references_len_diff = [abs(len(prediction) - length) for length in references_len_list]
-            min_index = references_len_diff.index(min(references_len_diff))
-            self.references_len += references_len_list[min_index]
-
+            self.references_len += len(
+                min(references, key=lambda x: abs(len(x) - self.pred_len)))
```
Comment (yhcc) [inline fastNLP/core/metrics/bleu.py]: 这里感觉应该是abs(len(x)-len(prediction))？

File: fastNLP/core/metrics/bleu.py [hunk 2]
```diff
@@ -25,68 +22,88 @@ def tokenizer_fn(sentence: str) -> Sequence[str]:
...
+               references: Sequence[Sequence[str]]) -> None:
         r"""
        :meth:`update` 函数将针对一个批次的预测结果做评价指标的累计。
-       :param predictions: 预测的 sentence, type为Sequence，长度可变，假设为L
-       :param references: 答案译文，type为Sequence，长度必须也为L，保持和predictions一致，每一个元素也是一个Sequence。
+       :param predictions: 预测的 ``sentence``, type为``Sequence``，长度可变，假设为 ``L``
+       :param references: 答案译文，type为``Sequence``，长度必须也为``L``，
+           保持和``predictions``一致，每一个元素也是一个``Sequence``。
        """
         # 为了兼容不同框架，我们将输入变量全部转为numpy类型来进行计算。
```
Comment (yhcc) [inline fastNLP/core/metrics/bleu.py]: 这一行删掉吧，有点误导，这里和numpy没有关系

File: fastNLP/core/metrics/bleu.py [hunk 3]
```diff
@@ -95,14 +112,10 @@ def update(self, predictions: Sequence[str], references: Sequence[Sequence[str]]
             counter_clip = pred_counter & reference_counter
 
             for counter in counter_clip:
```
Comment (yhcc) [inline fastNLP/core/metrics/bleu.py]: 这里counter修改一下名字，我理解此时循环出来的内容就是ngram吧，是不是写成for ngram in counter_clip会更加易懂一些。下面循环counter的也是。

File: fastNLP/core/metrics/bleu.py [hunk 4]
```diff
@@ -115,16 +128,18 @@ def get_metric(self) -> dict:
         precision_total = self.precision_total.tensor2numpy()
 
         if min(precision_matches) == 0.0:
-            return {"bleu": np.array(0.0)}
+            return {'bleu': np.array(0.0)}
         if self.smooth:
-            precision_score = np.add(precision_matches, np.ones(self.n_gram)) / np.add(precision_total,
-                                                                                       np.ones(self.n_gram))
+            precision_score = np.add(precision_matches, np.ones(
```
Comment (yhcc) [inline fastNLP/core/metrics/bleu.py]: 感觉这部分是不是直接写成precision_score = (precision_matches+1)/(precision_total+1)?

File: fastNLP/core/metrics/bleu.py [hunk 5]
```diff
@@ -115,16 +128,18 @@ def get_metric(self) -> dict:
...
+                self.n_gram)) / np.add(precision_total, np.ones(self.n_gram))
             precision_score[0] = precision_matches[0] / precision_total[0]
         else:
             precision_score = precision_matches / precision_total
 
-        precision_score = np.array(self.ngram_weights) * np.log(precision_score)
-        brevity_penalty = _get_brevity_penalty(self.pred_len.tensor2numpy(), self.references_len.tensor2numpy())
+        precision_score = np.array(
+            self.ngram_weights) * np.log(precision_score)
+        brevity_penalty = _get_brevity_penalty(
```
Comment (yhcc) [inline fastNLP/core/metrics/bleu.py:191]: 感觉应该是self.pred_len.to_scalar()?

File: fastNLP/core/metrics/element.py [hunk 6]
```diff
@@ -311,6 +315,18 @@ def __gt__(self, other) -> bool:
             other = other.value
         return self.value > other
 
+    def __getitem__(self, item):
+        assert isinstance(item, int)
+        if isinstance(self.init_value, Number):
+            raise ValueError(f"{type(self.init_value)} object does not support item assignment")
```
Comment (yhcc) [inline fastNLP/core/metrics/element.py]: 错误信息修改为TypeError: f'{type(self.init_value)}' object is not subscriptable

File: fastNLP/core/metrics/element.py [hunk 7]
```diff
@@ -311,6 +315,18 @@ def __gt__(self, other) -> bool:
...
+        else:
+            return self.to_list()[item]
+
+    def __setitem__(self, key, value):
+        if isinstance(self.init_value, Number):
+            raise ValueError(f"{type(self.init_value)} object does not support item assignment")
```
Comment (yhcc) [inline fastNLP/core/metrics/element.py]: 修改为TypeError

File: fastNLP/core/metrics/bleu.py [hunk 8]
```diff
@@ -25,84 +22,99 @@ def tokenizer_fn(sentence: str) -> Sequence[str]:
...
         self.smooth = smooth
         if ngram_weights is not None and len(ngram_weights) != n_gram:
-            raise ValueError(f"List of weights has different weights than `n_gram`: {len(ngram_weights)} != {n_gram}")
-        self.ngram_weights = ngram_weights if ngram_weights is not None else [1.0 / n_gram] * n_gram
-
-        self.register_element(name="pred_len", value=0, aggregate_method='sum', backend=backend)
-        self.register_element(name="references_len", value=0, aggregate_method='sum', backend=backend)
-        self.register_element(name="precision_matches", value=[0 for _ in range(self.n_gram)], aggregate_method='sum',
+            raise ValueError(
+                f'List of weights has different weights than `n_gram`: {len(ngram_weights)} != {n_gram}'
```
Comment (yhcc) [inline fastNLP/core/metrics/bleu.py]: raise ValueError(f'The number of weights in `weights` is different from `n_gram`: {len(ngram_weights)} != {n_gram}'

File: fastNLP/core/metrics/bleu.py [hunk 9]
```diff
@@ -21,88 +18,156 @@ def get_n_gram(token: Sequence[str], n_gram: int) -> Counter:
...
+    if lang == 'en':
+        return str.split
+    elif lang in ('cn', 'zh'):
+        return list
+    else:
+        return None
 
 
-def _get_brevity_penalty(pred_len: np.array, references_len: np.array) -> np.array:
+def _get_brevity_penalty(pred_len: np.array,
```
Comment (yhcc) [inline fastNLP/core/metrics/bleu.py]: 这个感觉不是返回一个float就好了， 为啥一定要np.array咧？

File: fastNLP/core/metrics/rouge.py [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+    """这个函数将根据命中数或者LCS和列表的长度去计算预测句子和标准译文句子的精度、召回率和F1得分值。
+
+    :param matches: 匹配数或最长公共子序列的长度。
+    :param pred_len: 预测句子的序列长度。
+    :param reference_len: 答案句子的序列长度。
+    """
+    precision = matches / pred_len
+    recall = matches / reference_len
+    if precision == recall == 0.0:
+        return dict(precision=np.array(0.0),
```
Comment (yhcc) [inline fastNLP/core/metrics/rouge.py]: 如果是一个数字的话，感觉没必要一定使用np.array

File: fastNLP/core/metrics/rouge.py [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+    :param pred: 一个预测句子的序列.
+    :param reference: 一个标准译文句子的序列.
+    """
+    pred_len, reference_len = len(pred), len(reference)
+    if 0 in (pred_len, reference_len):
+        return dict(precision=np.array(0.0),
+                    recall=np.array(0.0),
+                    fmeasure=np.array(0.0))
+    lcs = [[0] * (len(pred) + 1) for _ in range(len(reference) + 1)]
+    for i in range(1, len(reference) + 1):
```
Comment (yhcc) [inline fastNLP/core/metrics/rouge.py]: 看看能不能修改为SequenceMatcher的

File: fastNLP/core/metrics/rouge.py [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+                      for i in range(len(tokens) - n + 1)):
+            ngrams[ngram] += 1
+        return ngrams
+
+    pred_ngarms = get_n_gram(pred, n_gram)
+    reference_ngarms = get_n_gram(reference, n_gram)
+    pred_len = sum(pred_ngarms.values())
+    reference_len = sum(reference_ngarms.values())
+    if 0 in (pred_len, reference_len):
+        return dict(precision=np.array(0.0),
```
Comment (yhcc) [inline fastNLP/core/metrics/rouge.py]: 也是直接换成float吧

File: tests/core/metrics/test_element_cal_element.py [hunk 13]
```diff
@@ -25,6 +27,29 @@ def get_metric(self) -> dict:
...
+        self.c += pred
+        self.a /= 2
+        self.b -= self.c
+        self.c *= 2
+        self.b += 2
+        for i in range(4):
+            self.b[i] += i
+
+    def get_metric(self) -> dict:
+        return {"a": self.a.to_list(), "b": self.b.to_list(), "c": self.c.to_list()}
```
Comment (yhcc) [inline tests/core/metrics/test_element_cal_element.py]: 这里的self.c就不要to_list了吧，这样测试一下是否会自动to_list这样

File: fastNLP/core/metrics/rouge.py [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+                raise ValueError(
+                    f'Got unknown rouge key {rouge_key}. Expected to be one of {1 - 9} or L'  # noqa: E501
+                )
+        self.rouge_keys = rouge_keys
+        if use_stemmer:
+            if _module_available('nltk'):
+                import nltk
+                self.stemmer = nltk.stem.porter.PorterStemmer()
+            else:
+                raise ValueError('You need to download the nltk package')
```
Comment (yhcc) [inline fastNLP/core/metrics/rouge.py]: 报错信息：the nltk package is needed to use stemmer, check https://www.nltk.org/install.html for installation.

File: fastNLP/core/metrics/rouge.py [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+            reference_len = len(_references)
+            ref_tokens = []
+            for reference in _references:
+                ref_token = _normalize_and_tokenize(reference, self.stemmer,
+                                                    self.normalizer,
+                                                    self.tokenizer)
+                ref_tokens.append(ref_token)
+            for i, rouge_key in enumerate(self.rouge_keys):
+                fmeasure = precision = recall = 0
+                for j, reference in enumerate(_references):
```
Comment (yhcc) [inline fastNLP/core/metrics/rouge.py]: 这里就直接for ref_token in ref_tokens了吧，不然还得绕一下，还让reference这个东西用不上

File: fastNLP/core/metrics/rouge.py [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
+
+                self.fmeasure[i] += fmeasure
+                self.precision[i] += precision
+                self.recall[i] += recall
+
+    def get_metric(self) -> dict:
+        r"""
+        :meth:`get_metric` 函数将根据 :meth:`update` 函数累计的评价指标统计量来计算最终的评价结果。
+
+        :return: 包含以下内容的字典：``{ `rouge1_fmeasure` : float}``；
```
Comment (yhcc) [inline fastNLP/core/metrics/rouge.py]: 注释这里不是很准确，修改成：包含以下内容的字典：``{ 'rouge{x}_fmeasure': float, 'rouge{x}_precision': float, 'rouge{x}_recall': float}``，其中`{x}`的值取决于在初始化时传入的rouge_keys。

File: tests/core/metrics/test_rouge_torch.py [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+from fastNLP.envs.imports import _NEED_IMPORT_TORCH
+
+if _NEED_IMPORT_TORCH:
+    import torch
+    import torch.distributed
+    from torch.multiprocessing import Pool, set_start_method
+else:
+    from fastNLP.core.utils.dummy_class import DummyClass as set_start_method
+try:
+    from sklearn.metrics import accuracy_score as sklearn_accuracy
```
Comment (yhcc) [inline tests/core/metrics/test_rouge_torch.py]: 这个似乎没有用上？就没有必要import了哇？

File: fastNLP/core/metrics/rouge.py [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+                raise ValueError(
+                    f'Got unknown rouge key {rouge_key}. Expected to be one of {1 - 9} or L'  # noqa: E501
+                )
+        self.rouge_keys = rouge_keys
+        if use_stemmer:
+            if _module_available('nltk'):
+                import nltk
+                self.stemmer = nltk.stem.porter.PorterStemmer()
+            else:
+                raise ValueError('the nltk package is needed to use stemmer, check https://www.nltk.org/install.html for installation.')
```
Comment (yhcc) [inline fastNLP/core/metrics/rouge.py]: 开头的the首字母大写一下吧
